### PR TITLE
[Tabular] Online inference optimization

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- 2.5x+ speedup of online-inference latency in feature generation.
- 7.8x+ speedup when text features are present.
- ~30%+ speedup for FastAI NN inference when many features are present.
- Slight speedup for LightGBM when text features are present.

TODO:

- [x] Benchmark

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1136-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1136/3/index.html

**Comment by szha** [comment]
Job PR-1136-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1136/4/index.html

**Comment by Innixma** [comment]
Benchmark results are good, overall accuracy improvement, likely due to improved handling of rare categories (which also speeds up online-inference). Inference speedup isn't visible in the benchmark since it is batch inference. 0 errors.

```
AutoGluon_bestquality_1h_2021_04_24_reorder VS all
                                     framework  > AutoGluon_bestquality_1h_2021_04_24_reorder  < AutoGluon_bestquality_1h_2021_04_24_reorder  = AutoGluon_bestquality_1h_2021_04_24_reorder  % Less Avg. Errors Than AutoGluon_bestquality_1h_2021_04_24_reorder  time_train_s  metric_error  time_infer_s  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0   AutoGluon_bestquality_1h_2021_05_31_infopt                                            482                                            423                                             72                                            0.97036                     3350.043091  71991.170791     34.644831       0.444217               1.016989               1.264567  1.469806           482           495             0             0            0
1  AutoGluon_bestquality_1h_2021_04_24_reorder                                              0                                              0                                            977                                            0.00000                     3389.406448  72121.458546     33.485568       0.507677               1.027032               1.222007  1.530194           423           554             0             0            0
```

`AutoGluon_bestquality_1h_2021_04_24_reorder` is equivalent to mainline.
`AutoGluon_bestquality_1h_2021_05_31_infopt` is this PR's branch.

File: features/src/autogluon/features/generators/text_ngram.py [hunk 1]
```diff
@@ -194,13 +195,16 @@ def _generate_ngrams(self, X, downsample_ratio: int = None):
...
+                self._feature_names_dict[nlp_feature] = nlp_features_names_final
 
-            nlp_features_names = vectorizer_fit.get_feature_names()
-
-            X_nlp_features = pd.DataFrame(transform_matrix.toarray())  # TODO: Consider keeping sparse
-            X_nlp_features.columns = [f'{nlp_feature}.{x}' for x in nlp_features_names]
-            X_nlp_features[nlp_feature + '._total_'] = X_nlp_features.gt(0).sum(axis=1).astype(np.int32)
-
+            transform_array = transform_matrix.toarray()
+            nonzero_count = np.count_nonzero(transform_array, axis=1).astype(np.uint16)
```
Comment (gradientsky) [inline features/src/autogluon/features/generators/text_ngram.py:205]: Would uint16 max value a problem on the large datasets, where counts can overflow above 65535?

File: tabular/src/autogluon/tabular/augmentation/distill_utils.py [hunk 2]
```diff
@@ -71,10 +73,20 @@ def spunge_augment(X, feature_metadata: FeatureMetadata, num_augmented_samples=1
     if frac_perturb > 1.0:
         raise ValueError("frac_perturb must be <= 1")
     logger.log(20, f"SPUNGE: Augmenting training data with {num_augmented_samples} synthetic samples for distillation...")
+
+    X = X.copy()
+    category_featnames = feature_metadata.get_features(valid_raw_types=[R_CATEGORY])
+    for feature in category_featnames:
+        current_categories = X[feature].cat.categories
+        if '__NaN__' in current_categories:
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/augmentation/distill_utils.py:81]: nit: repeated use - consider using a constant to prevent defects on updates.

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py [hunk 3]
```diff
@@ -323,9 +319,9 @@ def _predict_proba(self, X, **kwargs):
         # fastai has issues predicting on a single row, duplicating the row as a workaround
         if single_row:
             X = pd.concat([X, X]).reset_index(drop=True)
-
         # Copy cat_columns and cont_columns because TabularList is mutating the list
-        test_dl = self.model.dls.test_dl(X)
+        # TODO: This call has very high fixed cost with many features (0.7s for a single row with 3k features)
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:323]: This is primarily due to normalization performed on the inputs.

File: features/src/autogluon/features/generators/text_ngram.py (hunk 1 above)
Comment (Innixma) [reply features/src/autogluon/features/generators/text_ngram.py:205]: Theoretically yes, but it would need to have 66k unique tokens present in a single row, which doesn't seem very likely to ever happen.

If it did happen, then the user has bigger problems to deal with than this feature overflowing (like the fact that the individual ngram counts are uint8, so if a single ngram appears > 511 times, it will overflow the counter).

Comment (Innixma) [reply features/src/autogluon/features/generators/text_ngram.py:206]: Added inline comment
